erpnext/regional/doctype/import_supplier_invoice/import_supplier_invoice.py

```diff
@@ -0,0 +1,386 @@
+			new_contact.first_name = args.supplier
+			new_contact.append('links', {
+				"link_doctype": "Supplier",
+				"link_name": existing_supplier_name
+			})
+			new_contact.insert()
+
+		return existing_supplier_name
+	else:
+		
+		new_supplier = frappe.new_doc("Supplier")
+		new_supplier.supplier_name = args.supplier
+		new_supplier.supplier_group = args.supplier_group
+		new_supplier.tax_id = args.tax_id
+		new_supplier.fiscal_code = args.fiscal_code
+		new_supplier.fiscal_regime = args.fiscal_regime
+		new_supplier.save()
+
+		new_contact = frappe.new_doc("Contact")
+		new_contact.first_name = args.supplier
+		new_contact.append('links', {
+			"link_doctype": "Supplier",
+			"link_name": new_supplier.name
+		})
+
+		new_contact.insert()
+
+		return new_supplier.name
+
+def create_address(**args):
+
+	args = frappe._dict(args)
+	filters = [
+			["Dynamic Link", "link_doctype", "=", "Supplier"],
+			["Dynamic Link", "link_name", "=", args.supplier_name],
+			["Dynamic Link", "parenttype", "=", "Address"]
+		]
+
+	existing_address = frappe.get_list("Address", filters)
+
+	if args.address_line1:
+		make_address = frappe.new_doc("Address")
+		make_address.address_line1 = args.address_line1
+
+		if args.city:
+			make_address.city = args.city
+		else:
+			make_address.city = "Not Provided"
+
+		if args.province:
+			make_address.state_code = args.province
+
+		if args.pincode:
+			make_address.pincode = args.pincode
+
+		if args.country:
+			make_address.country = args.country
+
+		for address in existing_address:
+			address_doc = frappe.get_doc("Address", address["name"])
+			if (address_doc.address_line1 == make_address.address_line1 and
+				address_doc.pincode == make_address.pincode):
+				return address
+
+		make_address.append("links", {
+			"link_doctype": "Supplier",
+			"link_name": args.supplier_name
+		})
+		make_address.address_type = "Billing"
+		make_address.insert()
+		return make_address.name
+	else:
+		return None
+
+def create_purchase_invoice(**args):
+
+	args = frappe._dict(args)
+	pi = frappe.get_doc({
+			"doctype": "Purchase Invoice",
+			"company": args.company,
+			"naming_series": args.naming_series,
+			"supplier": args.supplier_name,
+			"is_return": args.is_return,
+			"posting_date": today(),
+			"bill_no": args.bill_no,
+			"bill_date": args.bill_date,
+			"destination_code": args.destination_code,
+			"document_type": args.document_type,
+			"items": args["items"],
+			"taxes": args["taxes"]
+		})
+
+	try:
+		pi.insert(ignore_permissions=True)
+
+		if args.total_discount > 0:
+			pi.apply_discount_on = "Grand Total"
+			pi.discount_amount = args.total_discount
+			pi.save()
+
+		calc_total = 0
+		adj = 0
+		for term in args.terms:
+			calc_total += float(term["payment_amount"])
+		if float(calc_total - float(pi.grand_total)) != 0:
+			adj = calc_total - float(pi.grand_total)
+		pi.payment_schedule = []
+		for term in args.terms:
+			pi.append('payment_schedule',{"mode_of_payment_code": term["mode_of_payment_code"],
```

Comment:
Why can't append before insert?

Reply:
This is because grand total etc are calculated only after save